stan/math/gpu/matrix_gpu.hpp

```diff
@@ -90,23 +92,76 @@ class matrix_gpu {
       }
     }
   }
+
   /**
    * Constructor for the matrix_gpu that
-   * creates a copy of the Eigen matrix on the GPU.
+   * creates a copy of of the eigen matrix on the gpu and stores
+   *  the value in cache.
    *
+   * @tparam T type of matrix
+   * @tparam R rows of matrix
+   * @tparam C cols of matrix
+   * @param A Eigen matrix
    *
-   * @tparam T type of data in the Eigen matrix
+   * @throw <code>std::system_error</code> if the
+   * matrices do not have matching dimensions
+   */
+  template <typename T, int R, int C,
+            typename = typename std::enable_if<is_constant<T>::value, T>::type>
+  explicit matrix_gpu(const Eigen::Matrix<T, R, C>& A)
+      : rows_(A.rows()), cols_(A.cols()) {
+    cl::Context& ctx = opencl_context.context();
+    cl::CommandQueue& queue = opencl_context.queue();
+    if (A.opencl_buffer_() != NULL) {
+      oclBuffer_ = cl::Buffer(ctx, CL_MEM_READ_WRITE, sizeof(T) * A.size());
+      queue.enqueueCopyBuffer(A.opencl_buffer_, oclBuffer_, 0, 0,
+                              sizeof(T) * A.size());
+    } else {
+      if (size() > 0) {
+        try {
+          // creates the OpenCL buffer to copy the Eigen
+          // matrix to the OpenCL device
+          oclBuffer_ = cl::Buffer(ctx, CL_MEM_READ_WRITE, sizeof(T) * A.size());
+          A.opencl_buffer_
```

Comment:
Can this implementation call the other implementation above?

Reply:
Do you mean the implementation in copy?

Reply:
Yeah, I wonder if this can call copy and the var constructor can call this constructor with the result from `value_of`

Reply:
Let it be done!